networkx/drawing/layout.py

```diff
@@ -1121,6 +1122,236 @@ def multipartite_layout(G, subset_key="subset", align="vertical", scale=1, cente
+        Generate layout with or without considering the edge
+        weights
+    linlog : bool (default: False)
+        Use log attraction rather than linear attraction
+    dim: number (default: 2)
+       Sets the dimensions of the layout. This parameter is ignored if pos is given.
+
+    Examples
+    --------
+    >>> import networkx as nx
+    >>> G = nx.florentine_families_graph()
+    >>> nx.draw(G, pos=nx.forceatlas2_layout(G))
+
+    [1] https://journals.plos.org/plosone/article/file?id=10.1371/journal.pone.0098679&type=printable
+
+    """
+    import numpy as np
+
+    def estimate_factor(n, swing, traction, speed, speed_efficiency, jitter_tolerance):
+        """
+        ForceAtlas2 helper function
+        Computes scaling factor for force
+        """
+
+        # estimate jitter
+        opt_jitter = 0.05 * np.sqrt(n)
+        min_jitter = np.sqrt(opt_jitter)
+        max_jitter = 10
+        min_speed_efficiency = 0.05
+
+        other = min(max_jitter, opt_jitter * traction / n**2)
+        jitter = jitter_tolerance * max(min_jitter, other)
+
+        if swing / traction > 2.0:
+            if speed_efficiency > min_speed_efficiency:
+                speed_efficiency *= 0.5
+            jitter = max(jitter, jitter_tolerance)
+        if swing == 0:
+            target_speed = np.inf
+        else:
+            target_speed = jitter * speed_efficiency * traction / swing
+
+        if swing > jitter * traction:
+            if speed_efficiency > min_speed_efficiency:
+                speed_efficiency *= 0.7
+        elif speed < 1000:
+            speed_efficiency *= 1.3
+
+        max_rise = 0.5
+        speed = speed + min(target_speed - speed, max_rise * speed)
+        return speed, speed_efficiency
+
+    # parse optional pos positions
+    if pos is None:
+        pos = nx.random_layout(G, dim=dim)
+    else:
+        dim = len(next(iter(pos.values())))
+
+    # Scale the position of the node proportional to the @pos
+    max_pos_range = 1
+    min_pos_range = 0
+    # check if we have a valid pos else just return (empty graph)
+    if pos:
+        max_pos_range = np.array([max(i) for i in pos.values()]).max()
+        min_pos_range = np.array([min(i) for i in pos.values()]).min()
+    else:
+        return pos
+
+    # center locs around 0
+    pos_arr = (np.random.rand(len(G), dim) - 0.5) * (max_pos_range - min_pos_range)
+
+    mass = np.zeros(len(G))
+    size = np.zeros(len(G))
+    for idx, node in enumerate(G.nodes()):
+        if node in pos:
+            pos_arr[idx] = pos[node].copy()
+        mass[idx] = G.nodes[node].get("mass", G.degree(node) + 1)
+        size[idx] = G.nodes[node].get("size", 1)
+
+    n = G.number_of_nodes()
+    gravities = np.zeros((n, dim))
+    attraction = np.zeros((n, dim))
+    repulsion = np.zeros((n, dim))
+    weight = None
+    if edge_weight_influence:
+        weight = "weight"
+    A = nx.adjacency_matrix(G, weight=weight).todense()
```

Comment:
It's more efficient to create the dense adjacency directly rather than create the sparse adjacency matrix first then convert it to dense.

```suggestion
    A = nx.to_numpy_array(G, weight=weight)
```